// generation/dist/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function generator(source) {
  if (typeof source === 'function') {
    return source;
  } else if (typeof source[Symbol.iterator] === 'function') {
    return function* generate() {
      yield* source;
    };
  } else {
    throw new Error(`'${source}' must be either a generator function or an iterable`);
  }
}

class Generation {
  constructor(source = []) {
    Object.defineProperty(this, Symbol.iterator, {
      value: generator(source)
    });
  }

  get length() {
    let length = 0;
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = this[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        let x = _step.value;
        length++;
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator.return != null) {
          _iterator.return();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }

    return length;
  }

  equals(source) {
    return this.disjunct(source).length === 0;
  }

  map(fn) {
    return new Generation(function* () {
      var _iteratorNormalCompletion2 = true;
      var _didIteratorError2 = false;
      var _iteratorError2 = undefined;

      try {
        for (var _iterator2 = this[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
          let x = _step2.value;
          yield fn(x);
        }
      } catch (err) {
        _didIteratorError2 = true;
        _iteratorError2 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion2 && _iterator2.return != null) {
            _iterator2.return();
          }
        } finally {
          if (_didIteratorError2) {
            throw _iteratorError2;
          }
        }
      }
    }.bind(this));
  }

  filter(fn) {
    return new Generation(function* () {
      var _iteratorNormalCompletion3 = true;
      var _didIteratorError3 = false;
      var _iteratorError3 = undefined;

      try {
        for (var _iterator3 = this[Symbol.iterator](), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
          let item = _step3.value;

          if (fn(item)) {
            yield item;
          }
        }
      } catch (err) {
        _didIteratorError3 = true;
        _iteratorError3 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion3 && _iterator3.return != null) {
            _iterator3.return();
          }
        } finally {
          if (_didIteratorError3) {
            throw _iteratorError3;
          }
        }
      }
    }.bind(this));
  }

  concat(source) {
    return new Generation(function* () {
      yield* this;
      yield* new Generation(source);
    }.bind(this));
  }

  union(source) {
    return new Generation(function* () {
      let seen = new Set();
      var _iteratorNormalCompletion4 = true;
      var _didIteratorError4 = false;
      var _iteratorError4 = undefined;

      try {
        for (var _iterator4 = this[Symbol.iterator](), _step4; !(_iteratorNormalCompletion4 = (_step4 = _iterator4.next()).done); _iteratorNormalCompletion4 = true) {
          let x = _step4.value;

          if (!seen.has(x)) {
            seen.add(x);
            yield x;
          }
        }
      } catch (err) {
        _didIteratorError4 = true;
        _iteratorError4 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion4 && _iterator4.return != null) {
            _iterator4.return();
          }
        } finally {
          if (_didIteratorError4) {
            throw _iteratorError4;
          }
        }
      }

      var _iteratorNormalCompletion5 = true;
      var _didIteratorError5 = false;
      var _iteratorError5 = undefined;

      try {
        for (var _iterator5 = new Generation(source)[Symbol.iterator](), _step5; !(_iteratorNormalCompletion5 = (_step5 = _iterator5.next()).done); _iteratorNormalCompletion5 = true) {
          let y = _step5.value;

          if (!seen.has(y)) {
            seen.add(y);
            yield y;
          }
        }
      } catch (err) {
        _didIteratorError5 = true;
        _iteratorError5 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion5 && _iterator5.return != null) {
            _iterator5.return();
          }
        } finally {
          if (_didIteratorError5) {
            throw _iteratorError5;
          }
        }
      }
    }.bind(this));
  }

  intersect(source) {
    return new Generation(function* () {
      let seen = new Set();
      var _iteratorNormalCompletion6 = true;
      var _didIteratorError6 = false;
      var _iteratorError6 = undefined;

      try {
        for (var _iterator6 = new Generation(source)[Symbol.iterator](), _step6; !(_iteratorNormalCompletion6 = (_step6 = _iterator6.next()).done); _iteratorNormalCompletion6 = true) {
          let x = _step6.value;
          seen.add(x);
        }
      } catch (err) {
        _didIteratorError6 = true;
        _iteratorError6 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion6 && _iterator6.return != null) {
            _iterator6.return();
          }
        } finally {
          if (_didIteratorError6) {
            throw _iteratorError6;
          }
        }
      }

      yield* this.filter(y => seen.has(y));
    }.bind(this));
  }

  disjunct(source) {
    return new Generation(function* () {
      let seen = new Set();
      let disjunction = new Set();
      var _iteratorNormalCompletion7 = true;
      var _didIteratorError7 = false;
      var _iteratorError7 = undefined;

      try {
        for (var _iterator7 = this[Symbol.iterator](), _step7; !(_iteratorNormalCompletion7 = (_step7 = _iterator7.next()).done); _iteratorNormalCompletion7 = true) {
          let x = _step7.value;
          seen.add(x);
          disjunction.add(x);
        }
      } catch (err) {
        _didIteratorError7 = true;
        _iteratorError7 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion7 && _iterator7.return != null) {
            _iterator7.return();
          }
        } finally {
          if (_didIteratorError7) {
            throw _iteratorError7;
          }
        }
      }

      var _iteratorNormalCompletion8 = true;
      var _didIteratorError8 = false;
      var _iteratorError8 = undefined;

      try {
        for (var _iterator8 = new Generation(source)[Symbol.iterator](), _step8; !(_iteratorNormalCompletion8 = (_step8 = _iterator8.next()).done); _iteratorNormalCompletion8 = true) {
          let y = _step8.value;

          if (seen.has(y)) {
            disjunction.delete(y);
          } else {
            yield y;
          }
        }
      } catch (err) {
        _didIteratorError8 = true;
        _iteratorError8 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion8 && _iterator8.return != null) {
            _iterator8.return();
          }
        } finally {
          if (_didIteratorError8) {
            throw _iteratorError8;
          }
        }
      }

      yield* disjunction;
    }.bind(this));
  }

  reduce(fn, initial) {

    function reduce(fn, iterator, result) {
      let next = iterator.next();

      if (next.done) {
        return result;
      } else {
        return reduce(fn, iterator, fn(result, next.value));
      }
    }

    return reduce(fn, this[Symbol.iterator](), initial);
  }

}

function Generation$1(iterable) {
  return new Generation(iterable);
}
function map(iterable, fn) {
  return Generation$1(iterable).map(fn);
}
function filter(iterable, fn) {
  return Generation$1(iterable).filter(fn);
}
function concat(...iterables) {
  return iterables.reduce((generation, iterable) => {
    return generation.concat(iterable);
  }, Generation$1());
}
function union(...iterables) {
  return iterables.reduce((generation, iterable) => {
    return generation.union(iterable);
  }, Generation$1());
}
function intersect(first, ...iterables) {
  if (first) {
    return Generation$1(first).intersect(union(...iterables));
  } else {
    return Generation$1();
  }
}
function disjunct(...iterables) {
  return iterables.reduce((generation, iterable) => {
    return generation.disjunct(iterable);
  }, Generation$1());
}
function reduce(source, fn, initial) {
  return Generation$1(source).reduce(fn, initial);
}
function equals(left, right) {
  return Generation$1(left).equals(right);
}

exports.Generation = Generation$1;
exports.map = map;
exports.filter = filter;
exports.concat = concat;
exports.union = union;
exports.intersect = intersect;
exports.disjunct = disjunct;
exports.reduce = reduce;
exports.equals = equals;
exports.default = Generation$1;
